magento/magento2
# BUG - Prevent Magento from Creating incorrect URLs for category/products when Store View has modified URL keys

### Description (*)
If a Store View has a category tree where url_keys have been modified - eg different language - then if there is a category that has does not have a modified key (eg its the same in different languages) BUT has parents that DO, then the original code would use the DEFAULT Store View of the Category for creating URLS.

When scripts to regenerate URLs are used (as described in 28633) then they will end up creating incorrect URLs.

There is also a CORE issue. If an Admin changes the URL key of a category then Magento rebuilds all the URLs for the underlying categories and all the category/product URLs. However, if this is done where a category falls into the description above, then Magento sets the URL for that catalog/product combination for the store view it is done on but does it for ALL store views.

EG Given that the Default is English and there is a 3-level category :
Men > Collections > Premium 
(men) (collections) (premium) 

A product in Premium would have the url key : men/collections/premium/<product>

Now there are two other Store Views : Italian & German 
They have the names of the Men & Collections categories translated but not Premium as that is a brand.

So in the Italian Store the category should be -
(uomo) (collezioni) (premium)

German - 
(herren) (kollektionen) (premium)

IF an Admin went into the Italian store view and decided to change the URL key of 'uomo' to 'uomo-italiano' then Magento would rebuild all the keys for the view. If then you checked the url_redirect table for all the URLs for a single product for the "Premium" category then you would expect to see - 
men/collections/premium/<product>
herren/kollektionen/premium/<product>
uomo/collezioni/premium/<product>

BUT the bug causes all storeviews to be set to the Italian URL -
uomo/collezioni/premium/<product>
uomo/collezioni/premium/<product>
uomo/collezioni/premium/<product>

Categories where URL keys have been modified are okay.

### Related Pull Requests


### Fixed Issues (if relevant)

1. Fixes magento/magento2#28633

### Manual testing scenarios (*)

1. Set up categories as above
2. Change top level url key of a store view 
3. Check URLs of a category without a changed key for all store views - problem above shows.
4. Apply fix
5. Go to top level category in every store view and change key and revert 
6. Once all stores have been forced to update - check URLs for the category in 3 and all URLs for all stores will be correct.

### Questions or comments


### Contribution checklist (*)
